### PR TITLE
Fix vulnerability in a frame function

### LINKED ISSUE #136: `Frame::copy_from_raw_parts` can lead to segfault without `unsafe`
Hello :crab: ,
we (Rust group @sslab-gatech) found a memory-safety/soundness issue in this crate while scanning Rust code on crates.io for potential vulnerabilities.

## Issue Description
Since [`Frame::copy_from_raw_parts()`](https://docs.rs/av-data/0.2.2/av_data/frame/struct.Frame.html#method.copy_from_raw_parts) doesn't validate its input pointers,
it is possible to trigger undefined behavior without using `unsafe`.

## Proof of Concept
The example program below exhibits segmentation fault caused by dereferncing a null pointer.
```rust
use av_data::frame::*;
use av_data::pixel::*;

fn main() {
    let yuv420: Formaton = *formats::YUV420;
    let fm = std::sync::Arc::new(yuv420);
    let video_info = VideoInfo {
        pic_type: PictureType::I,
        width: 42,
        height: 42,
        format: fm,
    };
    let mut frame = new_default_frame(
        MediaKind::Video(video_info),
        None
    );

    frame.copy_from_raw_parts(
        vec![0 as usize as *const u8; 2].into_iter(), 
        vec![40; 2].into_iter()
    );

    println!("Program segfaults before reaching this point");
}
```

### Program Output (Ubuntu 18.04, rustc 1.50.0-nightly)
```
Segmentation fault (core dumped)
```

## Suggested Fix
`Frame::copy_from_raw_parts` API could be changed to an `unsafe` API,
since it requires users to check the validity of the input pointers in order to maintain safety guarantees.

Thank you for checking out this issue :+1:

### PR DESCRIPTION
**Pull request by Luni-4**
This PR fixes #136 

Thanks in advance for your review! :)

### CODE REVIEW

File: data/src/frame.rs [hunk 1]
```diff
@@ -388,7 +388,16 @@ impl Frame {
                 let s_linesize = src_linesize.next().unwrap();
                 let data = self.buf.as_mut_slice(i).unwrap();
                 let cc = f_iter.next().unwrap();
-                let rr = src.next().unwrap();
+                let rr = src
```
Comment (lu-zero) [inline data/src/frame.rs:391]: would be better zip() and panic if the source and the destination do not match.

**Review by lu-zero — CHANGES_REQUESTED** [review]

File: data/src/frame.rs (hunk 1 above)
Comment (Luni-4) [reply data/src/frame.rs:391]: Don't source and destination have a different memory location and different values? What should I compare?

**Comment by Qwaz** [comment]
This doesn't seem to fix the problem. Wouldn't it still segfaults with safe Rust if `0 as usize as *const u8` is changed to an arbitrary value (say `0x1234 as usize as *const u8`)?

**Comment by Luni-4** [comment]
Now it is really fixed!

Thanks @Qwaz for the help! :)

**Comment by Luni-4** [comment]
Strange, the test doesn't work on Windows....

**Review by lu-zero — APPROVED** [review]
Maybe we could add an additional assert.

**Comment by Luni-4** [comment]
> Maybe we could add an additional assert.

What do you propose?

**Comment by lu-zero** [comment]
Explicitly check for the plane count.

**Comment by Luni-4** [comment]
> Explicitly check for the plane count.

Sure, I could add that in another PR

**Pull request merged**